src/compiler/checker.ts

```diff
@@ -14960,12 +14960,10 @@ namespace ts {
                     }
                     // If no inferences can be made to K's constraint, infer from a union of the property types
                     // in the source to the template type X.
-                    const valueTypes = compact([
-                        getIndexTypeOfType(source, IndexKind.String),
-                        getIndexTypeOfType(source, IndexKind.Number),
-                        ...map(getPropertiesOfType(source), getTypeOfSymbol)
-                    ]);
-                    inferFromTypes(getUnionType(valueTypes), getTemplateTypeFromMappedType(target));
+                    const indexType = source.symbol && source.symbol.flags & SymbolFlags.Enum && getEnumKind(source.symbol) === EnumKind.Literal ?

```

Comment:
Rather than all this stuff, we shouldn't we just filter the `enumNumberIndexInfo` singleton from the result of `getIndexTypeOfType` so intersection-based inferences still work out?

Reply:
Yeah, that works.